python/cudf/cudf/core/join/join.py

```diff
@@ -0,0 +1,431 @@
+                    )
+                    self.rhs._data[rcol] = self.rhs._data[rcol].astype(
+                        to_dtype
+                    )
+        elif self.left_index and self.right_on:
+            for ((kl, vl), (kr, vr)) in zip(
+                self.lhs.index._data.items(), self.rhs._data.items()
+            ):
+                to_dtype = self.input_to_libcudf_casting_rules(
+                    vl, vr, self.how
+                )
+                self.lhs.index._data[kl] = vl.astype(to_dtype)
+                self.rhs._data[kr] = vr.astype(to_dtype)
+        elif self.right_index and self.left_on:
+            for ((kl, vl), (kr, vr)) in zip(
+                self.lhs._data.items(), self.rhs.index._data.items()
+            ):
+                to_dtype = self.input_to_libcudf_casting_rules(
+                    vl, vr, self.how
+                )
+                self.lhs._data[kl] = vl.astype(to_dtype)
+                self.rhs.index._data[kr] = vr.astype(to_dtype)
+
+        elif self.left_on and self.right_on:
+            for ((kl, vl), (kr, vr)) in zip(
+                self.lhs._data.items(), self.rhs._data.items()
+            ):
+                to_dtype = self.input_to_libcudf_casting_rules(
+                    vl, vr, self.how
+                )
+                self.lhs._data[kl] = vl.astype(to_dtype)
+                self.rhs._data[kr] = vr.astype(to_dtype)
+
+    def input_to_libcudf_casting_rules(self, lcol, rcol, how):
+        cast_warn = "can't safely cast column {} from {} with type \
+                        {} to {}, upcasting to {}"
+        ctgry_err = "can't implicitly cast column {0} to categories \
+                        from {1} during {1} join"
+
+        dtype_l = lcol.dtype
+        dtype_r = rcol.dtype
+        libcudf_join_type = None
+        if pd.api.types.is_dtype_equal(dtype_l, dtype_r):
+            # if categorical and equal, children passed to libcudf
+            libcudf_join_type = dtype_l
+        elif is_categorical_dtype(dtype_l) and is_categorical_dtype(dtype_r):
+            if how in ["inner", "outer"]:
+                # join as the underlying category dtype winner
+                libcudf_join_type = self.input_to_libcudf_casting_rules(
+                    dtype_l.categories, dtype_r.categories, "inner"
+                )
+            elif how == "left":
+                libcudf_join_type = self.input_to_libcudf_casting_rules(
+                    dtype_l.categories, dtype_r.categories, "left"
+                )
+
+        elif how == "left":
+            check_col = rcol.fillna(0)
+            if not check_col.can_cast_safely(dtype_l):
+                libcudf_join_type = self.input_to_libcudf_casting_rules(
+                    lcol, rcol, "inner"
+                )
+                warnings.warn(
+                    cast_warn.format(
+                        rcol, "right", dtype_r, dtype_l, libcudf_join_type
+                    )
+                )
```

Comment:
Why is the casting rule "inner" and the warn format "right" here?

Reply:
When we do a left join, we try and preserve the datatype on the left. But if the column on the right isn't overflow safe with respect to the left datatype, we fall back on the datatype we would have returned were it an inner join. This leads to `np.find_common_type` or just the max if they're the same kind. 

The warning would print as `cannot cast column *whatever* from the right to the left datatype`. This needs to change anyways though, because once `right` joins are actually just a reflection of the operands of a `left` join, we'll never reach the `right` branch of code. So it probably makes sense just to just refer to a general "column"

Reply:
Added